src/lib/concurrency/logging/group_commit_logger.cpp

```diff
@@ -0,0 +1,275 @@
+  entry.data.resize(entry.data.size() + number_of_bitmap_bytes);
+  auto null_bitmap_pos = entry.cursor;
+  entry.cursor += number_of_bitmap_bytes;
+
+  DebugAssert(entry.data[null_bitmap_pos] == '\0', "Logger: memory is not NULL");
+
+  EntryWriter visitor(entry);
+  auto bit_pos = 0u;
+  for (auto& value : values) {
+    auto has_content = boost::apply_visitor(visitor, value);
+
+    // Set corresponding bit in bitmap to 1 if the value is a NullValue
+    if (!has_content) {
+      entry.data[null_bitmap_pos] |= 0b00000001 << bit_pos;
+    }
+
+    // Increase bit_pos for next value and increase null_bitmap_pos every eigth values to adress the next byte
+    if (bit_pos == 7) {
+      ++null_bitmap_pos;
+      DebugAssert(entry.data[null_bitmap_pos] == '\0', "Logger: memory is not NULL");
+    }
+    bit_pos = (bit_pos + 1) % 8;
+  }
+
+  _write_to_buffer(entry.data);
+}
+
+void GroupCommitLogger::commit(const TransactionID transaction_id, std::function<void(TransactionID)> callback) {
+  constexpr auto entry_length = sizeof(char) + sizeof(TransactionID);
+  LogEntry entry(entry_length);
+
+  entry << 't' << transaction_id;
+
+  _commit_callbacks.emplace_back(std::make_pair(callback, transaction_id));
+
+  _write_to_buffer(entry.data);
+}
+
+void GroupCommitLogger::load_table(const std::string& file_path, const std::string& table_name) {
+  const auto entry_length = sizeof(char) + (file_path.size() + 1) + (table_name.size() + 1);
+  LogEntry entry(entry_length);
+
+  entry << 'l' << file_path << table_name;
+
+  _write_to_buffer(entry.data);
+}
+
+void GroupCommitLogger::invalidate(const TransactionID transaction_id, const std::string& table_name,
+                                   const RowID row_id) {
+  const auto entry_length =
+      sizeof(char) + sizeof(TransactionID) + (table_name.size() + 1) + sizeof(ChunkID) + sizeof(ChunkOffset);
+  LogEntry entry(entry_length);
+
+  entry << 'i' << transaction_id << table_name << row_id;
+
+  _write_to_buffer(entry.data);
+}
+
+void GroupCommitLogger::_write_to_buffer(std::vector<char>& entry) {
+  // Assume that there is always enough space in the buffer, since it is flushed on hitting half its capacity
+  DebugAssert(_buffer_position + entry.size() < _buffer_capacity, "logging: entry does not fit into buffer");
+
+  _buffer_mutex.lock();
+
+  memcpy(_buffer + _buffer_position, &entry[0], entry.size());
+
+  _buffer_position += entry.size();
+  _has_unflushed_buffer = true;
+
+  _buffer_mutex.unlock();
+
+  if (_buffer_position > _buffer_capacity / 2) {
+    flush();
+  }
+}
+
+void GroupCommitLogger::_write_buffer_to_logfile() {
+  DebugAssert(_log_file.is_open(), "Logger: Log file not open.");
+  _file_mutex.lock();
+  _buffer_mutex.lock();
+
+  _log_file.write(_buffer, _buffer_position);
+  _log_file.sync();
+
+  _buffer_position = 0u;
+  _has_unflushed_buffer = false;
+
+  for (auto& callback_tuple : _commit_callbacks) {
+    callback_tuple.first(callback_tuple.second);
```

Comment:
This does not look like something that should be the responsibility of a "write buffer to file" method.